rust-dap-rp2040/src/pio.rs

```diff
@@ -14,156 +14,255 @@
-
-        a.bind(&mut write_loop);
-        // Output 1-bit. and set SWCLK to High.
-        a.set(pio::SetDestination::PINS, HI);
-        a.out_with_delay(pio::OutDestination::PINS, 1, match Q { 0 => 0, _ => Q - 1 });
-        a.bind(&mut write_loop_enter);
-        // Keep looping unless X register is 0. and set SWCLK to Low.
-        a.jmp_with_delay_and_side_set(pio::JmpCondition::XDecNonZero, &mut write_loop, Q, LO);
-        // Set SWCLK to High.
-        a.set(pio::SetDestination::PINS, HI);
-        // Start over.
-        a.jmp(pio::JmpCondition::Always, &mut wrap_target);
-
-        // Read-Bits
-        a.bind(&mut read_start);
-        // Set IO direction of SWDIO pin to input.
-        a.out_with_delay(pio::OutDestination::PINDIRS, 1, match Q { 0 => 0, _ => Q - 1 });
-        // Jump if X register is not 0. with post decrement.
-        a.jmp(pio::JmpCondition::XDecNonZero, &mut read_loop_enter);
-        // Wait before reading SWDIO pin.
-        a.nop_with_delay(Q);
-        // Read state of SWDIO pin without clocking SWCLK.
-        a.r#in(pio::InSource::PINS, 1);
-        // Shift in 31 bits of 0 to MSB of ISR.
-        a.r#in(pio::InSource::NULL, 31);
-        // Put result data from ISR to FIFO.
-        a.push(false, true);
-        // Start over.
-        a.jmp(pio::JmpCondition::Always, &mut wrap_target);
-
-        a.bind(&mut read_loop);
-        // Shift-in 1-bit to ISR. and set SWCLK to High.
-        a.r#in_with_delay_and_side_set(pio::InSource::PINS, 1, Q, HI);
-        a.bind(&mut read_loop_enter);
-        // Keep looping unless X register is 0. and set SWCLK to Low.
-        a.jmp_with_delay_and_side_set(pio::JmpCondition::XDecNonZero, &mut read_loop, Q, LO);
-        a.r#in_with_side_set(pio::InSource::PINS, 1, HI);
-        // Put result data from ISR to FIFO.
-        a.push(false, true);
-        a.bind(&mut wrap_source);
-        // Start over.
-        // a.jmp(pio::JmpCondition::Always, &mut wrap_target);
-
-        // The labels wrap_target and wrap_source, as set above,
-        // define a loop which is executed repeatedly by the PIO
-        // state machine.
-        let program = a.assemble_with_wrap(wrap_source, wrap_target);
-        // let program = a.assemble_program();
-        // let program = program.set_origin(Some(0));
+        let program = all_pins_to_input_program();
 
         // Initialize and start PIO
+        // install swj_pin to pull-up RESET PIN
         let (mut pio, sm0, _, _, _) = pio0.split(resets);
         let installed = pio.install(&program).unwrap();
-        let (sm, rx, tx) = Self::build_pio(clk_pin_id, dat_pin_id, installed, DEFAULT_PIO_DIVISOR, sm0);
-        // Pin modes are controlled in connect() and disconnect() .
-        // sm.set_pindirs([(clk_pin_id, hal::pio::PinDir::Output), (dat_pin_id, hal::pio::PinDir::Output)]);
-
+        let (sm, rx, tx) = Self::build_pio((0,32) ,0, (0,32),0, installed, DEFAULT_PIO_DIVISOR, sm0);
```

Comment:
`SET` のピン数は `5` までなのでこのままだとpanicする。

```suggestion
        let (sm, rx, tx) = Self::build_pio((0,5) ,0, (0,32),0, installed, DEFAULT_PIO_DIVISOR, sm0);
```

Reply:
おっと、ありがとうございます。
これは私も確認していて直したはずでしたが、commitから漏れてたかもです。